compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirFunctionReturnTypeMismatch.kt

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.expression
+
+import org.jetbrains.kotlin.fir.analysis.checkers.compareTypesList
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors.RETURN_TYPE_MISMATCH
+import org.jetbrains.kotlin.fir.declarations.FirFunction
+import org.jetbrains.kotlin.fir.declarations.FirPropertyAccessor
+import org.jetbrains.kotlin.fir.declarations.FirSimpleFunction
+import org.jetbrains.kotlin.fir.expressions.FirReturnExpression
+import org.jetbrains.kotlin.fir.resolve.inference.isFunctionalType
+import org.jetbrains.kotlin.fir.typeContext
+import org.jetbrains.kotlin.fir.types.*
+import org.jetbrains.kotlin.types.AbstractTypeChecker
+
+object FirFunctionReturnTypeMismatch : FirReturnExpressionChecker() {
+    override fun check(expression: FirReturnExpression, context: CheckerContext, reporter: DiagnosticReporter) {
+        val returnExpressionType = expression.result.typeRef.coneTypeSafe<ConeKotlinType>() ?: return
+        val targetElement = expression.target.labeledElement
+
+        if (!(targetElement is FirSimpleFunction || targetElement is FirPropertyAccessor)) return
+
+        if (isNeedToBeReported(expression, targetElement, context.session.typeContext)) {
+            val returnExpressionSource = expression.source ?: return
+            val expectedConeType = targetElement.returnTypeRef.coneTypeSafe<ConeKotlinType>() ?: return
+            reporter.report(RETURN_TYPE_MISMATCH.on(returnExpressionSource, expectedConeType, returnExpressionType), context)
+        }
+    }
+
+    private fun isNeedToBeReported(
+        returnExpression: FirReturnExpression,
+        function: FirFunction<*>,
+        context: ConeInferenceContext
+    ): Boolean {
+        val functionReturnType = function.returnTypeRef.coneTypeSafe<ConeKotlinType>() ?: return false
+        val returnExpressionType = returnExpression.result.typeRef.coneTypeSafe<ConeKotlinType>() ?: return false
+        val functionLastTypeArg = functionReturnType.typeArguments.lastOrNull()?.type
+
+        if (functionReturnType.isFunctionalType(context.session) && functionLastTypeArg?.isUnit == true) {
```

Comment:
I'd try to extract this check or its part to some function and re-use it in FirInitializerTypeMismatchChecker.